diff --git a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiaaccessibility.cpp b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiaaccessibility.cpp
index 638d684ce4..5fae4408f6 100644
--- a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiaaccessibility.cpp
+++ b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiaaccessibility.cpp
@@ -142,6 +142,9 @@ void QWindowsUiaAccessibility::notifyAccessibilityUpdate(QAccessibleEvent *event
     case QAccessible::NameChanged:
         QWindowsUiaMainProvider::notifyNameChange(event);
         break;
+    case QAccessible::IdentifierChanged:
+        QWindowsUiaMainProvider::notifyIdentifierChange(event);
+        break;
     case QAccessible::SelectionAdd:
         QWindowsUiaMainProvider::notifySelectionChange(event);
         break;
diff --git a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
index 56181caa21..ddb0e7b6b5 100644
--- a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
+++ b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
@@ -173,6 +173,18 @@ void QWindowsUiaMainProvider::notifyNameChange(QAccessibleEvent *event)
     }
 }
 
+void QWindowsUiaMainProvider::notifyIdentifierChange(QAccessibleEvent *event)
+{
+    if (QAccessibleInterface *accessible = event->accessibleInterface()) {
+        if (auto provider = providerForAccessible(accessible)) {
+            QComVariant oldVal;
+            QComVariant newVal{ accessible->text(QAccessible::Identifier) };
+            UiaRaiseAutomationPropertyChangedEvent(provider.Get(), UIA_AutomationIdPropertyId,
+                                                   oldVal.get(), newVal.get());
+        }
+    }
+}
+
 void QWindowsUiaMainProvider::notifySelectionChange(QAccessibleEvent *event)
 {
     if (QAccessibleInterface *accessible = event->accessibleInterface()) {
diff --git a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.h b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.h
index 03954d2805..decd381183 100644
--- a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.h
+++ b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.h
@@ -33,6 +33,7 @@ public:
     static void notifyStateChange(QAccessibleStateChangeEvent *event);
     static void notifyValueChange(QAccessibleValueChangeEvent *event);
     static void notifyNameChange(QAccessibleEvent *event);
+    static void notifyIdentifierChange(QAccessibleEvent *event);
     static void notifySelectionChange(QAccessibleEvent *event);
     static void notifyTextChange(QAccessibleEvent *event);
     static void raiseNotification(QAccessibleAnnouncementEvent *event);
diff --git a/qtbase/src/widgets/accessible/qaccessiblewidget.cpp b/qtbase/src/widgets/accessible/qaccessiblewidget.cpp
index b0bb12ea86..3960f825af 100644
--- a/qtbase/src/widgets/accessible/qaccessiblewidget.cpp
+++ b/qtbase/src/widgets/accessible/qaccessiblewidget.cpp
@@ -393,6 +393,8 @@ QString QAccessibleWidget::text(QAccessible::Text t) const
         break;
     case QAccessible::Value:
         break;
+    case QAccessible::Identifier:
+        str = widget()->accessibleId();
     default:
         break;
     }
diff --git a/qtbase/src/widgets/kernel/qwidget.cpp b/qtbase/src/widgets/kernel/qwidget.cpp
index 881c8ba9a4..a14ec736af 100644
--- a/qtbase/src/widgets/kernel/qwidget.cpp
+++ b/qtbase/src/widgets/kernel/qwidget.cpp
@@ -11859,6 +11859,20 @@ QString QWidget::accessibleDescription() const
     Q_D(const QWidget);
     return d->accessibleDescription;
 }
+
+void QWidget::setAccessibleId(const QString &id)
+{
+    Q_D(QWidget);
+    d->accessibleId = id;
+    QAccessibleEvent event(this, QAccessible::IdentifierChanged);
+    QAccessible::updateAccessibility(&event);
+}
+
+QString QWidget::accessibleId() const
+{
+    Q_D(const QWidget);
+    return d->accessibleId;
+}
 #endif // QT_CONFIG(accessibility)
 
 #ifndef QT_NO_SHORTCUT
diff --git a/qtbase/src/widgets/kernel/qwidget.h b/qtbase/src/widgets/kernel/qwidget.h
index 647640a1fe..7130608010 100644
--- a/qtbase/src/widgets/kernel/qwidget.h
+++ b/qtbase/src/widgets/kernel/qwidget.h
@@ -166,6 +166,7 @@ class Q_WIDGETS_EXPORT QWidget : public QObject, public QPaintDevice
 #if QT_CONFIG(accessibility)
     Q_PROPERTY(QString accessibleName READ accessibleName WRITE setAccessibleName)
     Q_PROPERTY(QString accessibleDescription READ accessibleDescription WRITE setAccessibleDescription)
+    Q_PROPERTY(QString accessibleId READ accessibleId WRITE setAccessibleId)
 #endif
     Q_PROPERTY(Qt::LayoutDirection layoutDirection READ layoutDirection WRITE setLayoutDirection RESET unsetLayoutDirection)
     QDOC_PROPERTY(Qt::WindowFlags windowFlags READ windowFlags WRITE setWindowFlags)
@@ -406,6 +407,8 @@ public:
     void setAccessibleName(const QString &name);
     QString accessibleDescription() const;
     void setAccessibleDescription(const QString &description);
+    QString accessibleId() const;
+    void setAccessibleId(const QString &id);
 #endif
 
     void setLayoutDirection(Qt::LayoutDirection direction);
diff --git a/qtbase/src/widgets/kernel/qwidget_p.h b/qtbase/src/widgets/kernel/qwidget_p.h
index 9016532ba2..caa9989d92 100644
--- a/qtbase/src/widgets/kernel/qwidget_p.h
+++ b/qtbase/src/widgets/kernel/qwidget_p.h
@@ -683,6 +683,7 @@ public:
 #if QT_CONFIG(accessibility)
     QString accessibleName;
     QString accessibleDescription;
+    QString accessibleId;
 #endif
 
     // Other variables.
